# Machine learning prediction and experimental validation of antigenic drift in H3 influenza A viruses in swine

Zeller, M.A., Gauger, P.C., Arendsee, Z.W., Souza, C.K., Vincent, A.L., and Anderson, T.K. Machine learning prediction and experimental validation of antigenic drift in H3 influenza A viruses in swine. bioRxiv Preprint doi: TBD.

Abstract: The genetic and antigenic diversity of influenza A virus (IAV) circulating in swine challenges the development of effective vaccines, thereby increasing the zoonotic threat and pandemic potential of swine IAV. High throughput sequencing technologies and analyses are able to quantify genetic diversity of IAV, but there are no accurate approaches to adequately describe novel antigenic phenotypes. This study evaluated an ensemble of non-linear regression models to estimate virus phenotype from genotype. Regression models were trained with a phenotypic dataset of pairwise hemagglutination inhibition (HI) assays, using genetic sequence identity and pairwise amino acid mutations as predictor features. The model identified pairwise amino acid identity, ranked the relative importance of mutations in the hemagglutinin (HA) protein, and demonstrated good prediction accuracy following ten-fold cross validation. Four previously untested IAV strains were selected to experimentally validate the model predictions by HI assays. Error between predicted and measured distances of uncharacterized strains were 0.34, 0.70, 2.19, and 0.17 antigenic units. These regression models trained on HI data can be used to estimate antigenic distances between different strains of IAV in swine using sequence data. By ranking the importance of mutations in the HA, this method provides criteria for identifying antigenically advanced IAV strains that may not be controlled by existing vaccines and can inform strain updates to vaccines to better control this important pathogen.

